amy/consents/tests/test_action_required_view.py

```diff
@@ -0,0 +1,298 @@
+            publish_profile=True,
+            data_privacy_agreement=True,
+        )
+        self.form_url = reverse("action_required_terms")
+
+    def test_anonymous_user(self):
+        """Ensure anonymous user is not redirected by the Terms middleware."""
+        urls = [
+            reverse("login"),
+            reverse("api:root"),
+            reverse("training_request"),
+            reverse("training_request_confirm"),
+            reverse("workshop_request"),
+            reverse("workshop_request_confirm"),
+        ]
+        # ensure we're not logged in
+        self.client.logout()
+
+        for url in urls:
+            rv = self.client.get(url)
+            # no redirects!
+            self.assertEqual(rv.status_code, 200)
+            # user indeed is anonymous
+            self.assertEqual(rv.wsgi_request.user.is_anonymous, True)
+
+    def test_logged_in_user(self):
+        """Ensure logged-in user who has not consented to
+        the required terms is redirected to the form."""
+        urls = [
+            reverse("admin-dashboard"),
+            reverse("trainee-dashboard"),
+        ]
+
+        # ensure we're logged in
+        self.client.force_login(self.neville)
+        # ensure we have not yet agreed to the required consents
+        self.assertEqual(person_has_consented_to_required_terms(self.neville), False)
+        with self.terms_middleware():
+            for url in urls:
+                rv = self.client.get(url)
+                # redirects to the form
+                self.assertEqual(rv.status_code, 302)
+                self.assertTrue(rv["Location"].startswith(self.form_url))
+
+    def test_no_more_redirects_after_agreement(self):
+        """Ensure user is no longer forcefully redirected to accept the
+        required terms."""
+        url = reverse("trainee-dashboard")
+
+        # ensure we're logged in
+        self.client.force_login(self.neville)
+        self.assertEqual(person_has_consented_to_required_terms(self.neville), False)
+
+        with self.terms_middleware():
+            # we can't get to the url because we're redirected to the form
+            rv = self.client.get(url)
+            self.assertEqual(rv.status_code, 302)
+            self.assertTrue(rv["Location"].startswith(self.form_url))
+
+            # agree on the required terms
+            self.person_agree_to_terms(
+                self.neville,
+                Term.objects.filter(required_type=Term.PROFILE_REQUIRE_TYPE),
+            )
+
+            # now the dashboard is easily reachable
+            rv = self.client.get(url)
+            self.assertEqual(rv.status_code, 200)
+
+    def test_allowed_urls(self):
+        urls = [
+            reverse("logout"),
+        ]
+        # ensure we're logged in
+        self.client.force_login(self.neville)
+        self.assertEqual(person_has_consented_to_required_terms(self.neville), False)
+        with self.terms_middleware():
+            for url in urls:
```

Comment:
You probably want to force_login in the loop, in case there are more URLs to cover.